src/drivers/s3/S3Disk.ts

```diff
@@ -0,0 +1,502 @@
+            }
+            throw error;
+        }
+        if (params.Key.endsWith('/')) {
+            throw new NotAFileError(path);
+        }
+        return this.s3Client.getObject(params).createReadStream();
+    }
+
+    /**
+     * Determines if a provided object in a bucket is a directory explicitly (ends in a `/`) or
+     * a directory object already exists with the same name excluding the `/`.
+     *
+     * @param params
+     */
+    private async doObjectParamsMatchDirectoryObject(
+        params: S3ObjectParams,
+    ): Promise<boolean> {
+        if (params.Key.endsWith('/')) {
+            return true;
+        }
+        /*
+         * We also want to check to see if a key with the same name exists with a `/` at the end
+         * and disallow that as well since it would be equivelant on a normal filesystem (whereas
+         * it's acceptable in S3 since the `/` is part of the object key and thus `foo` and `foo/`
+         * are considered separate objects.
+         */
+        try {
+            await this.s3Client
+                .headObject({
+                    Bucket: params.Bucket,
+                    Key: `${params.Key}/`,
+                })
+                .promise();
+            /*
+             * If we get to this point without throwing, it means a directory object exists so
+             * we indicate as much.
+             */
+            return true;
+        } catch (error) {
+            if (error.code !== 'NotFound') {
+                /*
+                 * NotFound error is good! It means a directory object doesn't exist with the same
+                 * name as the provided key. We re-throw on any other error.
+                 */
+                throw error;
+            }
+        }
+        /*
+         * If we get to this point, it means the headObject threw but it was just an expected
+         * NotFound error. So that means we can consider the provided key a writable path.
+         */
+        return false;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public async write(
+        path: string,
+        body: string | Buffer | Readable,
+    ): Promise<void> {
+        const params = this.getPutObjectParams(path);
+        if (await this.doObjectParamsMatchDirectoryObject(params)) {
+            throw new NotWritableDestinationError(path);
+        }
+
+        // If we got a Readable stream we need to use `upload` instead of `putObject` since `upload` can handle
+        // streams of arbitrary length.
+        const operation =
+            typeof body === 'object' && body instanceof Readable
+                ? this.s3Client.upload({ ...params, Body: body })
+                : this.s3Client.putObject({ ...params, Body: body });
+
+        await operation.promise();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public async createWriteStream(): Promise<Writable> {
+        // TODO Fudge this by returning a Stream wrapper around the putObject stream functionality.
```

Comment:
Is this TODO still relevant, do we need a task somewhere to cover it?

Reply:
Yep still relevant but nowhere to track it. This isn't crucial to functionality and just a note to a future developer who wants to implement this feature.